### ISSUE TITLE
Error causing an indexing pipeline crash 

### ISSUE THREAD

**Issue by fmassot** [quickwit-oss/quickwit#2247]
Cluster is running on k8s with a 1 `metastore`, 1 `janitor`, 1 `indexer` and 1 `searcher`.   

I'm particularly worried about this log `quickwit-indexer-0 quickwit thread 'non-blocking-0' panicked at 'Cannot start a runtime from within a runtime. `

On the index, I have these logs:

```
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z ERROR quickwit_actors::spawn_builder: actor-failure cause=Failed to receive command from uploader.
quickwit-indexer-0 quickwit
quickwit-indexer-0 quickwit Caused by:
quickwit-indexer-0 quickwit     channel closed exit_status=Failure(Failed to receive command from uploader.
quickwit-indexer-0 quickwit
quickwit-indexer-0 quickwit Caused by:
quickwit-indexer-0 quickwit     channel closed)
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-blue-LT9X exit_status=Failure(cause=Failed to receive command from uploader.
quickwit-indexer-0 quickwit
quickwit-indexer-0 quickwit Caused by:
quickwit-indexer-0 quickwit     channel closed)
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z ERROR quickwit_actors::actor: actor-failure actor_name="quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-blue-LT9X" actor_exit_status=Failure(Failed to receive command from uploader.
quickwit-indexer-0 quickwit
quickwit-indexer-0 quickwit Caused by:
quickwit-indexer-0 quickwit     channel closed)
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z ERROR quickwit_actors::actor: exit activating-kill-switch actor=quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-blue-LT9X exit_status=Failure(Failed to receive command from uploader.
quickwit-indexer-0 quickwit
quickwit-indexer-0 quickwit Caused by:
quickwit-indexer-0 quickwit     channel closed)
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=Publisher-frosty-99K3 exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.007Z ERROR quickwit_actors::actor: actor-failure actor_name="Publisher-frosty-99K3" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=Indexer-throbbing-Hk4a exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z ERROR quickwit_actors::actor: actor-failure actor_name="Indexer-throbbing-Hk4a" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=quickwit_indexing::actors::doc_processor::DocProcessor-falling-9ci3 exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z ERROR quickwit_actors::actor: actor-failure actor_name="quickwit_indexing::actors::doc_processor::DocProcessor-falling-9ci3" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=SourceActor-little-SHF6 exit_status=Downstream actor exited.
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z ERROR quickwit_actors::actor: actor-failure actor_name="SourceActor-little-SHF6" actor_exit_status=DownstreamClosed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.102Z ERROR quickwit_actors::actor: exit activating-kill-switch actor=SourceActor-little-SHF6 exit_status=DownstreamClosed
quickwit-indexer-0 quickwit thread 'non-blocking-0' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', quickwit-indexing/src/source/kafka_source.rs:153:22
quickwit-indexer-0 quickwit note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=quickwit_indexing::actors::index_serializer::IndexSerializer-cool-jhK3 exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z ERROR quickwit_actors::actor: actor-failure actor_name="quickwit_indexing::actors::index_serializer::IndexSerializer-cool-jhK3" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=Packager-winter-yxup exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z ERROR quickwit_actors::actor: actor-failure actor_name="Packager-winter-yxup" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z  INFO quickwit_actors::spawn_builder: actor-exit actor_id=Uploader-snowy-L7iF exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.163Z ERROR quickwit_actors::actor: actor-failure actor_name="Uploader-snowy-L7iF" actor_exit_status=Killed
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="SourceActor-little-SHF6"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="quickwit_indexing::actors::doc_processor::DocProcessor-falling-9ci3"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="Indexer-throbbing-Hk4a"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="quickwit_indexing::actors::index_serializer::IndexSerializer-cool-jhK3"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="Packager-winter-yxup"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="Uploader-snowy-L7iF"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-blue-LT9X"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_actors::actor_handle: actor-exit-without-success actor="Publisher-frosty-99K3"
quickwit-indexer-0 quickwit 2022-11-03T18:21:03.874Z ERROR quickwit_indexing::actors::indexing_pipeline: Indexing pipeline failure. pipeline_id=IndexingPipelineId { index_id: "quickwit-fts-staging", source_id: "kafka-source", node_id: "quickwit-indexer-0", pipeline_ord: 0 } generation=13 healthy_actors=[] failed_or_unhealthy_actors=["SourceActor-little-SHF6", "quickwit_indexing::actors::doc_processor::DocProcessor-falling-9ci3", "Indexer-throbbing-Hk4a", "quickwit_indexing::actors::index_serializer::IndexSerializer-cool-jhK3", "Packager-winter-yxup", "Uploader-snowy-L7iF", "quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-blue-LT9X", "Publisher-frosty-99K3"] success_actors=[]
quickwit-indexer-0 quickwit 2022-11-03T18:21:08.524Z ERROR quickwit_actors::actor_handle: actor-timeout actor="IndexingPipeline-wispy-XLhD"
quickwit-indexer-0 quickwit 2022-11-03T18:21:08.524Z ERROR quickwit_indexing::actors::indexing_service: Indexing pipeline failed. index_id=quickwit-fts-staging source_id=kafka-source pipeline_ord=0
```

**Comment by fulmicoton** [comment]
You truncated too late. Can you give earlier logs? The error suggests your uploader had a problem.

**Comment by fulmicoton** [comment]
By pipeline crash, you mean the indexing did not resume  after this, correct?

**Comment by fulmicoton** [comment]
@guilload let's just use flume for that channel.

**Comment by fmassot** [comment]
yes the indexing pipeline is not resumed, the only log you can see after is coming from the ingest API garbage collector.

It's not the same logs as above but same issue triggered this time by an Azure storage misconfiguration (I copy/paste some logs directly from logz and wasn't able to download the real logs).

```


Nov 3, 2022 @ 20:08:32.881 | [2m2022-11-03T19:08:32.881Z[0m [32m INFO[0m [2mquickwit_indexing::actors::ingest_api_garbage_collector[0m[2m:[0m ingest-api-garbage-collect-operation
-- | --

  | Nov 3, 2022 @ 20:02:08.038 | [2m2022-11-03T19:02:08.037Z[0m [31mERROR[0m [2mquickwit_indexing::actors::indexing_service[0m[2m:[0m Indexing pipeline failed. [3mindex_id[0m[2m=[0mquickwit-fts-staging [3msource_id[0m[2m=[0mkafka-source [3mpipeline_ord[0m[2m=[0m0

  | Nov 3, 2022 @ 20:02:08.038 | [2m2022-11-03T19:02:08.037Z[0m [31mERROR[0m [2mquickwit_actors::actor_handle[0m[2m:[0m actor-timeout [3mactor[0m[2m=[0m"IndexingPipeline-empty-RG5g"

  | Nov 3, 2022 @ 20:02:04.089 | [2m2022-11-03T19:02:04.088Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"Publisher-blue-JbqQ" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.089 | channel closed)

  | Nov 3, 2022 @ 20:02:04.089 | Caused by:

  | Nov 3, 2022 @ 20:02:04.089 | (empty)

  | Nov 3, 2022 @ 20:02:04.089 | [2m2022-11-03T19:02:04.088Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"quickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-spring-WefF" [3mactor_exit_status[0m[2m=[0mFailure(Failed to receive command from uploader.

  | Nov 3, 2022 @ 20:02:04.089 | [2m2022-11-03T19:02:04.088Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mPublisher-blue-JbqQ [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.089 | (empty)

  | Nov 3, 2022 @ 20:02:04.089 | [2m2022-11-03T19:02:04.088Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m exit activating-kill-switch [3mactor[0m[2m=[0mquickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-spring-WefF [3mexit_status[0m[2m=[0mFailure(Failed to receive command from uploader.

  | Nov 3, 2022 @ 20:02:04.089 | channel closed)

  | Nov 3, 2022 @ 20:02:04.089 | Caused by:

  | Nov 3, 2022 @ 20:02:04.088 | Caused by:

  | Nov 3, 2022 @ 20:02:04.088 | (empty)

  | Nov 3, 2022 @ 20:02:04.088 | [2m2022-11-03T19:02:04.088Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mquickwit_indexing::actors::sequencer::Sequencer<quickwit_indexing::actors::publisher::Publisher>-spring-WefF [3mexit_status[0m[2m=[0mFailure(cause=Failed to receive command from uploader.

  | Nov 3, 2022 @ 20:02:04.088 | (empty)

  | Nov 3, 2022 @ 20:02:04.088 | Caused by:

  | Nov 3, 2022 @ 20:02:04.088 | (empty)

  | Nov 3, 2022 @ 20:02:04.088 | channel closed)

  | Nov 3, 2022 @ 20:02:04.088 | channel closed)

  | Nov 3, 2022 @ 20:02:04.088 | Caused by:

  | Nov 3, 2022 @ 20:02:04.088 | channel closed [3mexit_status[0m[2m=[0mFailure(Failed to receive command from uploader.

  | Nov 3, 2022 @ 20:02:04.088 | [2m2022-11-03T19:02:04.088Z[0m [31mERROR[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-failure [3mcause[0m[2m=[0mFailed to receive command from uploader.

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.066Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"IndexUploader-floral-tIcY" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.066Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mIndexUploader-floral-tIcY [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.066Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"Packager-divine-6iRj" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.065Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"quickwit_indexing::actors::index_serializer::IndexSerializer-holy-lF84" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.065Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mquickwit_indexing::actors::index_serializer::IndexSerializer-holy-lF84 [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.066 | [2m2022-11-03T19:02:04.066Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mPackager-divine-6iRj [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.021 | note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

  | Nov 3, 2022 @ 20:02:04.021 | thread 'non-blocking-0' panicked at 'Cannot start a runtime from within a runtime. This happens because a function (like `block_on`) attempted to block the current thread while the thread is being used to drive asynchronous tasks.', quickwit-indexing/src/source/kafka_source.rs:153:22

  | Nov 3, 2022 @ 20:02:04.012 | [2m2022-11-03T19:02:04.012Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"quickwit_indexing::actors::doc_processor::DocProcessor-bold-rN0W" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.012 | [2m2022-11-03T19:02:04.012Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mquickwit_indexing::actors::doc_processor::DocProcessor-bold-rN0W [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.011 | [2m2022-11-03T19:02:04.011Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"SourceActor-fragrant-ZEPr" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.011 | [2m2022-11-03T19:02:04.011Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mSourceActor-fragrant-ZEPr [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.011 | [2m2022-11-03T19:02:04.011Z[0m [31mERROR[0m [2mquickwit_actors::actor[0m[2m:[0m actor-failure [3mactor_name[0m[2m=[0m"Indexer-floral-bsRk" [3mactor_exit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.011 | [2m2022-11-03T19:02:04.011Z[0m [32m INFO[0m [2mquickwit_actors::spawn_builder[0m[2m:[0m actor-exit [3mactor_id[0m[2m=[0mIndexer-floral-bsRk [3mexit_status[0m[2m=[0mKilled

  | Nov 3, 2022 @ 20:02:04.002 | 1: AzureErrorWrapper(inner=server returned error status which will not be retried: 400) [3msplit_id[0m[2m=[0m"01GGZDWAXM6JY1WHMYP1SDJK9Y"

  | Nov 3, 2022 @ 20:02:04.002 | Caused by:

  | Nov 3, 2022 @ 20:02:04.002 | (empty)

  | Nov 3, 2022 @ 20:02:04.002 | [2m2022-11-03T19:02:04.002Z[0m [33m WARN[0m [1mindex_batch[0m[1m{[0m[3mindex_id[0m[2m=[0mquickwit-fts-staging [3msource_id[0m[2m=[0mkafka-source [3mpipeline_ord[0m[2m=[0m0[1m}[0m[2m:[0m[1muploader[0m[2m:[0m [2mquickwit_indexing::actors::uploader[0m[2m:[0m Failed to upload split. Killing! [3mcause[0m[2m=[0mFailed uploading key 01GGZDWAXM6JY1WHMYP1SDJK9Y.split in bucket azure://quickwit-eastus-staging@quickwit-account/quickwit/quickwit-fts-staging

  | Nov 3, 2022 @ 20:02:04.002 | 0: StorageError(kind=Service, source=AzureErrorWrapper(inner=server returned error status which will not be retried: 400))



```

**Comment by fulmicoton** [comment]
> @guilload Can we use a Mailbox, Inbox for that channel maybe?
>
> Then we can use the unbounded high priority channel for rebalances.

Answering myself: This may not be a panacea.
I actually don't answer why unti test / rebalancing works usually but here it didn't.

I first assumed that it was some logic that made it so that the local runtime was only started when try_send failed, but apparently not. Adrien, do you know something about this function is called outside of tokio usually but in some corner case, can be called from an async function?

**Comment by fulmicoton** [comment]
@fmassot Ah ok. So the uploader bug is nothing to worry about. It is just a misconfiguration. Understood.

**Comment by guilload** [comment]
`subscribe` is called in the constructor, that's the issue:
```cpp
 * @remark subscribe() is an asynchronous method which returns immediately:
 *         background threads will (re)join the group, wait for group rebalance,
 *         issue any registered rebalance_cb, assign() the assigned partitions,
 *         and then start fetching messages. This cycle may take up to
 *         \c session.timeout.ms * 2 or more to complete.
 ```

**Comment by guilload** [comment]
I'm going to move the subscribing logic in the blocking consumer poll loop.

**Comment by guilload** [comment]
Closed via #2248.
